model/task/task.go

```diff
@@ -2486,82 +2487,89 @@ func (t *Task) Insert() error {
+// that need to be (accounting for ResetFailedWhenFinished) in two queries
 func ArchiveMany(tasks []Task) error {
-	if len(tasks) == 0 {
-		return nil
-	}
-	existingTasksMap := map[string]bool{}
-	for _, t := range tasks {
-		existingTasksMap[t.Id] = true
-	}
-	additionalTasks := []string{}
+	bundledTasks := []string{}        // Contains all tasks and display tasks
+	execTasks := []string{}           // Contains all exec tasks
+	toUpdateExecTaskIds := []string{} // Contains all exec tasks that should update and have new execution
+	toArchive := []interface{}{}      // Contains all archived tasks that are being updated / have new execution
+
 	for _, t := range tasks {
-		// For any display tasks here, make sure that we also archive all their execution tasks.
-		// Consider each execution tasks individually, so we don't query a task we already passed in.
-		for _, et := range t.ExecutionTasks {
-			if !existingTasksMap[et] {
-				additionalTasks = append(additionalTasks, et)
-				existingTasksMap[et] = true
+		bundledTasks = append(bundledTasks, t.Id)
+		toArchive = append(toArchive, t.makeArchivedTask())
+		if t.DisplayOnly && len(t.ExecutionTasks) > 0 {
+			eTasks, err := FindAll(db.Query(ByIds(t.ExecutionTasks)))
+			if err != nil {
+				return errors.Wrapf(err, "finding execution tasks for display task '%s'.", t.Id)
+			}
+			for _, et := range eTasks {
+				execTasks = append(execTasks, et.Id)
+				if shouldRestartTask(t.ResetFailedWhenFinished, t.ResetWhenFinished, et.Status) {
+					toArchive = append(toArchive, et.makeArchivedTask())
+					toUpdateExecTaskIds = append(toUpdateExecTaskIds, et.Id)
+				}
 			}
 		}
 	}
-	if len(additionalTasks) > 0 {
-		toAdd, err := FindAll(db.Query(ByIds(additionalTasks)))
-		if err != nil {
-			return errors.Wrap(err, "finding execution tasks")
-		}
-		tasks = append(tasks, toAdd...)
-	}
 
-	archived := []interface{}{}
-	taskIds := []string{}
-	for _, t := range tasks {
-		archived = append(archived, *t.makeArchivedTask())
-		taskIds = append(taskIds, t.Id)
-	}
-	grip.DebugWhen(len(utility.UniqueStrings(taskIds)) != len(taskIds), message.Fields{
+	grip.DebugWhen(len(utility.UniqueStrings(bundledTasks)) != len(bundledTasks), message.Fields{
 		"ticket":           "EVG-17261",
 		"message":          "archiving same task multiple times",
-		"tasks_to_archive": taskIds,
+		"tasks_to_archive": bundledTasks,
 	})
 
+	err := archiveAll(bundledTasks, execTasks, toUpdateExecTaskIds, toArchive)
+
+	return err
+}
+
+// archiveAll takes in:
+// - taskIds         : All tasks id's of display tasks, execution tasks, and tasks
+// - toUpdateTaskIds : All tasks that should be progressing to the latest execution
+// - toArchive       : Compiled archived tasks to put in old tasks collection
+//
+// Each array *should* include execution tasks id's
+func archiveAll(taskIds, execTasksIds, toUpdateExecTaskIds []string, toArchive []interface{}) error {
```

Comment:
Can we make these a bit clearer? It seems to me like taskIds is all tasks that are being reset, since we update execution. toUpdateExecTaskIds isn't very clear since we also run an update on execTaskIds. Seems more like those are the exec tasks being reset. Also I wonder if, in the case that toUpdateExecTaskIds == execTaskIds, we could have a combined update instead of split updates.

cc @bynn for that last idea we could split a new ticket, up to you

Reply:
I forgot to update the comments... I just did and I think they're really clear now and consistent with other comments (please lmk if it is cleaner)

The combined update would be possible with some checks- my only concern with it is that it will be a large query. Should I implement the grand query?

Reply:
The query would be the same and only the update would be different right? I think a split ticket is better since this is a large code review

Reply:
I envision it as

if execTask == toUpdateExecTask {
// Combined query
} else {
// Existing query
}

But the combined query will be similar to the first PR version I had with a large query (but with bson.M instead of bson.D). The combined query would look like
$in: ids
$set: LPE = !LPE ? Execution + 1 : LPE + 1 
$set: Execution = LPE (so we get the newest LPE)
$unset --

The combined query could replace both by adding one pipeline
$in: execTaskIds
$set: LPE = !LPE ? Execution + 1 : LPE + 1 
$in: toUpdateExecTaskIds
$set: Execution = LPE (so we get the newest LPE)
$unset --

and we only have one query